python/shotgun_data/shotgun_data_retriever.py

```diff
@@ -679,23 +867,44 @@ def _get_thumbnail_path(url, bundle, directory_only=False):
         # sharding methodology, see 
         # http://stackoverflow.com/questions/13841931/using-guids-as-folder-names-splitting-up
         #
-        # From the hash, generate paths on the form C1C2/C3C4/rest_of_hash.jpeg
-        # (where C1 is the first character of the hash.)
-        # for a million evenly distributed items, this means ~15 items per folder
+        # From the hash, generate paths on the form C1C2/C3C4/rest_of_hash
+        # (where C1 is the first character of the hash). For a million evenly distributed
+        # items, this means ~15 items per folder.
         first_folder = hash_str[0:2]
         second_folder = hash_str[2:4]
-        path_chunks = [first_folder, second_folder]
 
+        # Establish the cache path directory
+        cache_path_items = [
+            bundle.cache_location, "thumbs", first_folder, second_folder
+        ]
+
+        cached_thumb_exists = False
         # If we were only asked to give back a directory path then we can
         # skip building and appending a file name.
         if not directory_only:
-            path_chunks.append("%s.jpeg" % hash_str[4:])
-
-        # establish the root path
-        cache_path_items = [bundle.cache_location, "thumbs"]
-        # append the folders
-        cache_path_items.extend(path_chunks)
-        # join up the path
+            # Look for an existing cache file. Use the glob module since
+            # we do not know what the file type of the cache file is.
+            path_base = hash_str[4:]
+            cache_base = os.path.join(*(cache_path_items + [path_base]))
+
+            # Attempt to match something that looks like:
+            #   /bundle_cache_location/thumbs/C1C2/C3C4/rest_of_hash.*
+            cache_matches = glob.glob("%s.*" % cache_base)
+            if len(cache_matches):
```

Comment:
not sure if it's overkill, but maybe worth throwing in a warning if there is more than 1 match?